File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/intake/Extend.java
package org.firstinspires.ftc.teamcode.intake;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Intake.Extend.in;
import static org.firstinspires.ftc.teamcode.Positions.Intake.Extend.out;

/**
 * Class representing the Extend mechanism of the robot.
 * It extends the ServoPositionMechanism class to control the extension servos.
 */
public class Extend extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servos from.
     */
    public Extend(HardwareMap hardwareMap) {
        super(in);
        this.hardwareMap = hardwareMap;
    }
    /**
     * Gets the servos associated with the Extend mechanism.
     * @return an array of servos.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Extend1"), hardwareMap.get(Servo.class, "Extend2")};
    }
}
/**
 * TeleOp class for testing the Extend mechanism manually.
 */
@TeleOp(name = "Intake Extend Test", group = "C Intake")
class ExtendTest extends ManualMechanismTeleOp {
    public ExtendTest() {
        super(Extend::new);
    }
}
/**
 * TeleOp class for testing the positions of the Extend mechanism.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Intake Extend Positions Test", group = "D Intake")
@Config
class ExtendPositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Extend extend = new Extend(hardwareMap);

        waitForStart();

        while (opModeIsActive()) {
            if (gamepad1.dpad_up) {
                current = in;
            } else if (gamepad1.dpad_down) {
                current = out;
            } else {
                current += extend.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            }
            extend.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if (!extend.run(packet)) requestOpModeStop();
            FtcDashboard.getInstance().sendTelemetryPacket(packet);
            telemetry.addData("Position of Extend", extend.getTargetPosition());
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/intake/Extend.kt
package org.firstinspires.ftc.intake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.intake.Extend.Companion.`in`
import org.firstinspires.ftc.intake.Extend.Companion.out
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.ServoPositionMechanism

/**
 * The Extend mechanism of the intake system.
 *
 * @param hardwareMap The hardware map for the robot.
 */
@Config
class Extend(hardwareMap: HardwareMap) : ServoPositionMechanism(`in`) {
    companion object {
        /**
         * The in position of the extend.
         */
        @JvmField
        @Volatile
        var `in` = 0.635
        /**
         * The out position of the extend.
         */
        @JvmField
        @Volatile
        var out = 0.9
    }
    /**
     * Gets the servo associated with the Extend mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Intake Extend"])
}
/**
 * TeleOp class for testing the Extend mechanism manually.
 */
@TeleOp(name = "Extend Test", group = "C")
@Disabled
private class ExtendTest : ManualMechanismTeleOp(::Extend)

/**
 * TeleOp class for testing the positions of the Extend mechanism.
 */
@TeleOp(name = "Extend Position Test", group = "D")
@Disabled
private class ExtendPositions : Movement() {
    /**
     * The Extend mechanism to be tested.
     */
    override lateinit var system: Extend

    override fun systemInit() {
        system = Extend(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_up) { (system as Extend).targetPosition = `in` })
        actions.add(PressAction(gamepad1::dpad_down) { (system as Extend).targetPosition = out })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}